zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -574,35 +574,43 @@ mod tests {
                 fee INTEGER,
                 FOREIGN KEY (block) REFERENCES blocks(height)
             )",
-            "CREATE TABLE transparent_received_output_spends (
+            r#"CREATE TABLE "transparent_received_output_spends" (
                 transparent_received_output_id INTEGER NOT NULL,
                 transaction_id INTEGER NOT NULL,
                 FOREIGN KEY (transparent_received_output_id)
-                    REFERENCES utxos(id)
+                    REFERENCES transparent_received_outputs(id)
                     ON DELETE CASCADE,
                 FOREIGN KEY (transaction_id)
                     -- We do not delete transactions, so this does not cascade
                     REFERENCES transactions(id_tx),
                 UNIQUE (transparent_received_output_id, transaction_id)
+            )"#,
+            "CREATE TABLE transparent_received_outputs (
+                id INTEGER PRIMARY KEY,
+                transaction_id INTEGER,
+                output_index INTEGER NOT NULL,
```

Comment:
`transaction_id` is nullable but `output_index` is not. Do we ever receive a transparent output without having its transaction?

In the `utxos` table we had `prevout_txid BLOB NOT NULL`, but we have no equivalent of that here. So if it were the case that we can see a transparent output with knowledge of only the outpoint and no corresponding row in `transactions`, this would be a loss of information relative to `utxos`.

Also, I don't think that `CONSTRAINT transparent_output_unique UNIQUE (transaction_id, output_index)` does the intended thing if `transaction_id` can be null.

Reply:
`transaction_id` should not be nullable, this is an oversight.

Reply:
@daira and I noted that the act of fetching UTXOs _is_ scanning for a wallet, so we should require the caller to always call `getTransaction` for each UTXO afterwards. That means we will always have the full transaction to insert into the wallet.

We do also need to be able to set `received_height`, which exists as a temporary crutch until we can rework the `blocks` table to not only contain fully-scanned blocks, so rather than directly using `decrypt_and_store_tx` we may want to rework `put_utxo` to instead take the full transaction.

Reply:
Whenever we call `put_legacy_transparent_utxo` we have the tx_ref (i.e. `transaction_id`) anyway, so we can just add it as a parameter to that call. Then `transaction_id` can be non-nullable.

Reply:
Oh! `WalletTransparentOutput` should include `transaction_id`. We only construct that type in these situations:
* from a row of the `utxos` / `transparent_received_outputs` table (in `to_unspent_transparent_output`);
* in `store_sent_tx` for ephemeral outputs (after #1257), in which case we have `transaction_id` because we've just obtained it from `wallet::put_tx_data`;
* in tests where we could easily make a fake transaction (which would improve the realism of the test).

Reply:
@str4d and/or @nuttycom pointed out that `wallet::put_legacy_transparent_utxo` is also called from `wallet::put_received_transparent_utxo` which is exposed via the data API. (E.g. see [Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_putUtxo](https://github.com/Electric-Coin-Company/zcash-android-wallet-sdk/blob/159de09b2bf1cc1c752260c76d444c11dfabebc9/backend-lib/src/main/rust/lib.rs#L1518) in the Android SDK.)

In that case, we should obtain the txid from the `WalletTransparentOutput`'s `output.hash`, and use it to look up the transaction in case we have it. If we don't have it, then we might have to synthesize a new entry in `transactions` with mostly null fields. However, we do always know the height in that case, because we created the `WalletTransparentOutput` from a `GetAddressUtxosReply` which has the height. So we either create the transaction entry with a known height, or we upsert that height to the existing entry.

If, on the other hand, we are adding an entry to `transparent_received_outputs` in order to record the ephemeral output of the first transaction in a ZIP 320 pair (once #1257 has merged), then we have the tx_ref so can record that immediately. We will only have the height once the transaction is mined (if it is), and so joining on `transactions` is the correct way to get the height for that output at any point in time.

In neither case do we need a `received_height` field directly in `transparent_received_outputs`, and it violates normalization to have it.

Reply:
We can't set the height of the transaction unless we've already scanned the corresponding block, because the height is a foreign key to the `blocks` table (which uses the block height as its primary key.).

Reply:
For completeness: I suggested adding a `mined_height` field to `transactions`, but adding a constraint that it is equal to `block` when the latter is non-null. This avoids problems with a height recorded in `transparent_received_outputs` potentially being inconsistent between outputs in the same transaction. The PR currently takes my suggested approach, and `transaction_id` is no longer nullable, so this is resolved.